src/main/java/build/buildfarm/instance/shard/ShardInstance.java

```diff
@@ -2104,6 +2105,58 @@ public ListenableFuture<Void> queue(ExecuteEntry executeEntry, Poller poller, Du
         operationTransformService);
   }
 
+  private ListenableFuture<Action> fetchAction(
+      RequestMetadata requestMetadata, ExecuteOperationMetadata metadata, Operation operation) {
+    Digest actionDigest = metadata.getActionDigest();
+
+    log.log(
+        Level.FINE,
+        format(
+            "ShardInstance(%s): queue(%s): fetching action %s",
+            getName(), operation.getName(), actionDigest.getHash()));
+
+    return FluentFuture.from(expectAction(actionDigest, requestMetadata))
+        .transformAsync(
+            action -> transformActionViaActionCache(action, metadata, operation),
+            operationTransformService)
+        .catchingAsync(
+            StatusException.class,
+            e -> describeFailedActionFetch(e, actionDigest),
+            operationTransformService);
+  }
+
+  private ListenableFuture<Action> transformActionViaActionCache(
+      Action action, ExecuteOperationMetadata metadata, Operation operation)
+      throws IOException, StatusException {
+    if (action == null) {
+      throw Status.NOT_FOUND.asException();
+    }
+
+    if (action.getDoNotCache()) {
+      // invalidate our action cache result as well as watcher owner
+      actionCache.invalidate(DigestUtil.asActionKey(metadata.getActionDigest()));
+      backplane.putOperation(
+          operation.toBuilder().setMetadata(Any.pack(action)).build(), metadata.getStage());
+    }
+
+    return immediateFuture(action);
+  }
+
+  private ListenableFuture<Action> describeFailedActionFetch(Throwable e, Digest actionDigest)
+      throws StatusException {
+    Status st = Status.fromThrowable(e);
+    if (st.getCode() == Code.NOT_FOUND) {
+      PreconditionFailure.Builder preconditionFailure = PreconditionFailure.newBuilder();
+      preconditionFailure
+          .addViolationsBuilder()
+          .setType(VIOLATION_TYPE_MISSING)
+          .setSubject("blobs/" + DigestUtil.toString(actionDigest))
+          .setDescription(MISSING_ACTION);
+      checkPreconditionFailure(actionDigest, preconditionFailure.build());
+    }
+    throw st.asRuntimeException();
+  }
```

Comment:
This framing of the exception filtering in a method makes this pattern look pretty awkward. Is there a clearer/cleaner way to express what we're dealing with here? The if block here is a noescape/noreturn, and we should not be throwing asRuntimeException from a Status when we declare to throw StatusException (asException() preferred).